File: frontend/src/api/push.ts
import { apiClient } from './client'

export async function getVapidKey(): Promise<string> {
  const { data } = await apiClient.get<{ public_key: string }>('/api/v1/push/vapid-key')
  return data.public_key
}

export async function subscribePush(subscription: PushSubscription): Promise<void> {
  const json = subscription.toJSON()
  await apiClient.post('/api/v1/push/subscribe', {
    endpoint: json.endpoint,
    keys: {
      p256dh: json.keys?.p256dh,
      auth: json.keys?.auth,
    },
  })
}

export async function unsubscribePush(endpoint: string): Promise<void> {
  await apiClient.delete('/api/v1/push/subscribe', {
    data: { endpoint },
  })
}

File: backend/app/api/push.py
from fastapi import APIRouter, Depends
from sqlalchemy.ext.asyncio import AsyncSession

from app.api.dependencies import get_current_user
from app.domain.entities import User
from app.infrastructure.database import get_session
from app.infrastructure.repositories.push_repository import PushRepository
from app.infrastructure.vapid_manager import get_vapid_public_key
from app.schemas.push import SubscribeRequest, SubscribeResponse, UnsubscribeRequest, VapidKeyResponse

router = APIRouter()


@router.get("/vapid-key", response_model=VapidKeyResponse)
async def vapid_key(current_user: User = Depends(get_current_user)):
    return VapidKeyResponse(public_key=get_vapid_public_key())


@router.post("/subscribe", response_model=SubscribeResponse, status_code=201)
async def subscribe(
    body: SubscribeRequest,
    current_user: User = Depends(get_current_user),
    session: AsyncSession = Depends(get_session),
):
    repo = PushRepository(session)
    is_new = await repo.save_subscription(
        user_id=current_user.id,
        endpoint=body.endpoint,
        p256dh_key=body.keys.p256dh,
        auth_key=body.keys.auth,
    )
    status = "subscribed" if is_new else "already_subscribed"
    return SubscribeResponse(status=status)


@router.delete("/subscribe", status_code=204)
async def unsubscribe(
    body: UnsubscribeRequest,
    current_user: User = Depends(get_current_user),
    session: AsyncSession = Depends(get_session),
):
    repo = PushRepository(session)
    await repo.delete_subscription_by_endpoint(body.endpoint)
